elastic/beats
# Add a persistent cache for cloudfoundry metadata

## What does this PR do?

Add a persistent cache on disk for Cloudfoundry metadata, based on the statestore.

## Why is it important?

Improve restart of beats when monitoring very big Cloudfoundry deployments.

## Checklist


## Author's Checklist

- [ ] Review performance, make obvious optimizations.
- [ ] Review encoding used to store structures.
   - [ ] Why custom labels are not being used when marshalling to JSON?
- [ ] Cache should be closed at some moment.
- [ ] Review error handling when key is missing vs. actual error.
- [ ] Add settings to make it optional?
- [ ] Use different store names for different Cloudfoundry deployments, to support monitoring multiple clusters from the same Beat (maybe not needed because it is difficult to have uuid collisions).
- [ ] When ready, separate in two PRs, one for the persistent cache, and another one for the change in `add_cloudfoundry_metadata`.

## How to test this PR locally

* Use `add_cloudfoundry_metadata` with Filebeat or Metricbeat.
* Restart the Beat.
* The Beat should reuse information on disk to enrich events without querying cloudfoundry API.

## Related issues

- Would replace #20707, where I evaluate the use of Auditbeat datastore for the same purpouse.
- Fixes #19511.
